src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/activities/ActivitiesService.kt

```diff
@@ -0,0 +1,117 @@
+import uk.gov.justice.digital.hmpps.prisonertonomisupdate.services.PayRateRequest
+import java.math.BigDecimal
+import java.time.LocalDateTime
+
+@Service
+class ActivitiesService(
+  private val activitiesApiService: ActivitiesApiService,
+  private val nomisApiService: NomisApiService,
+  private val mappingService: ActivitiesMappingService,
+  private val activitiesUpdateQueueService: ActivitiesUpdateQueueService,
+  private val telemetryClient: TelemetryClient
+) {
+
+  private companion object {
+    val log: Logger = LoggerFactory.getLogger(this::class.java)
+  }
+
+  fun createActivity(event: OutboundHMPPSDomainEvent) {
+    activitiesApiService.getActivitySchedule(event.identifier).run {
+      val activity = activitiesApiService.getActivity(this.activity.id)
+
+      val telemetryMap = mutableMapOf(
+        "activityScheduleId" to id.toString(),
+        "description" to description,
+      )
+
+      // to protect against repeated create messages for same activity
+      if (mappingService.getMappingGivenActivityScheduleId(event.identifier) != null) {
+        log.warn("Mapping already exists for activity schedule id ${event.identifier}")
+        return
+      }
+
+      val nomisResponse = try {
+        nomisApiService.createActivity(toNomisActivity(this, activity))
+      } catch (e: Exception) {
+        telemetryClient.trackEvent("activity-create-failed", telemetryMap)
+        log.error("createActivity() Unexpected exception", e)
+        throw e
+      }
+
+      val mapWithNomisId = telemetryMap
+        .plus(Pair("courseActivityId", nomisResponse.courseActivityId.toString()))
+
+      try {
+        mappingService.createMapping(
+          ActivityMappingDto(
+            nomisCourseActivityId = nomisResponse.courseActivityId,
+            activityScheduleId = event.identifier,
+            mappingType = "ACTIVITY_CREATED",
+          )
+        )
+      } catch (e: Exception) {
+        telemetryClient.trackEvent("activity-create-map-failed", mapWithNomisId)
+        log.error("Unexpected exception, queueing retry", e)
+        activitiesUpdateQueueService.sendMessage(
+          ActivityContext(
+            nomisCourseActivityId = nomisResponse.courseActivityId,
+            activityScheduleId = event.identifier,
+          )
+        )
+        return
+      }
+
+      telemetryClient.trackEvent("activity-created-event", mapWithNomisId)
+    }
+  }
+
+  private fun toNomisActivity(schedule: ActivitySchedule, activity: Activity): CreateActivityRequest {
+    return CreateActivityRequest(
+      code = "$activity.id-$schedule.id",
+      startDate = activity.startDate,
+      endDate = activity.endDate,
+      prisonId = activity.prisonCode,
+      internalLocationId = schedule.internalLocation!!.id.toLong(),
+      capacity = schedule.capacity,
+      payRates = activity.pay.map { p ->
+        PayRateRequest(
+          incentiveLevel = p.incentiveLevel!!,
```

Comment:
maybe for the future - would it be possible for the activities team to make these fields mandatory if they really are required